### PR TITLE
tests, sriov: Do not mount /sys/devices/ for SRIOV devices

### PR DESCRIPTION
**Pull request by alonSadan**
**What this PR does / why we need it**:
Libvirt required a Read/Write access to the hostdev devices representation,
on sysfs in older versions (< 5.6).
Since libvirt 5.6.0-10 [1] this is no longer the case and Read-Only access is sufficient.
Therefore, there is no longer a need for kubevirt to explicitly mount
/sys/devices with RW access on the virt-launcher.

This change removes the sysfs /sys/devices mount for VMI/s which contain SRIOV interfaces.

It does not change the logic for VMI/s which contain GPU devices,
as those need to be examined and tested separately to see that no degradation exists.
A following change should handle GPU/s.

1.https://bugzilla.redhat.com/show_bug.cgi?id=1758330

Signed-off-by: alonSadan <asadan@redhat.com>

colaborated with @ormergi  on this one.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
None
```

### CODE REVIEW

File: tests/vmi_multus_test.go
Comment (oshoval) [inline tests/vmi_multus_test.go:837]: do we really need this test ?
all the tests create VM, in case of such fundamental problem, we would see the error and i am not sure we want to verify on each and every time that the directory doesnt exists?

File: pkg/virt-controller/services/template.go
Comment (oshoval) [inline pkg/virt-controller/services/template.go:360]: for GPU we still need ?

File: go.mod
Comment (oshoval) [inline go.mod:73]: u need or a leftover?

**Comment by oshoval** [comment]
didnt you say that there cil file has an entry that need to be removed?

File: tests/vmi_multus_test.go
Comment (alonSadan) [reply tests/vmi_multus_test.go:837]: well, I thought of it of way of proving the change actually does what we think it does. I wasn't sure it is needed either.
I can drop it if you prefer.

File: pkg/virt-controller/services/template.go
Comment (alonSadan) [reply pkg/virt-controller/services/template.go:360]: Not sure. but out of the scope of this PR I Think.

File: go.mod
Comment (alonSadan) [reply go.mod:73]: I am not sure. I deleted it and it created it again.

**Comment by alonSadan** [comment]
> didnt you say that the cil file has an entry that need to be removed?
@oshoval I am not sure I understand what you were saying here. Can you clarify?

File: tests/vmi_multus_test.go
Comment (oshoval) [reply tests/vmi_multus_test.go:837]: IMO, we shouldnt check it, if we remove a mount, its not there, we should see that VM works and e2e pass thats all.
Maybe worth to mention the libvirt version where it was first not needed?

File: pkg/virt-controller/services/template.go
Comment (oshoval) [reply pkg/virt-controller/services/template.go:360]: sure np, but wonder if we want to address it in the future

File: go.mod
Comment (oshoval) [reply go.mod:73]: i would suggest you to remove just the mount, and run the generate, if you see it still add it, you will know its something in your env,

the only thing that can create it is the test or your local env, from a glance its not the tests (but i might be wrong because i just glanced)

**Comment by oshoval** [comment]
> > didnt you say that the cil file has an entry that need to be removed?
> 
> @oshoval I am not sure I understand what you were saying here. Can you clarify?

You and Or shown this file
https://github.com/kubevirt/kubevirt/blob/master/cmd/virt-handler/virt_launcher.cil#L57
```
    ; Allowing libvirtd to write to /sys files.
    ; This permission is necessary when using SR-IOV, but is not needed for anything else.
    (allow process sysfs_t (file (write)))
```
Dont you want to remove the write permission ? 
since you remove the W mount, and you are with R only

**Review by EdDev — COMMENTED** [review]
Thank you! Great news.

See inline comments and below about the commit.

>  sriov: remove host mounts for virt-launcher pods

tests, sriov: Do not mount `/sys/devices/` for SRIOV devices

> In the past, for sriov-vms, libvirt needed to open the `config` file under /sys/devices in
> read-write mode. This was not possible since k8s mounted it as read only.
> To overcome this sys/devices/ was mounted to the sriov virt-launcher pods in
> read-write mode. Since then libvirt was patched to not attempt to open
> the `config` file in read-write mode, and the whole /sys/* is mounted
> as read-only to the virt-launcher pods. For that, there is no need for
> the mount of /sys/devices anymore.

Libvirt required a Read/Write access to the hostdev devices representation on sysfs in older versions (< 5.6). Since libvirt 5.6.0-10 [1] this is no longer the case and Read(Only) access is sufficient.

Therefore, there is no longer a need for kubevirt to explicitly mount `/sys/devices` with RW access on the virt-launcher.

> This commit removes the mount of /sys/devices/ completely, modifies the unit test
> and adds a functional test, to verify it.

This change removes the sysfs mount for VMI/s which contain SRIOV interfaces.

It does not change the logic for VMI/s which contain GPU devices as those need to be examined and tested separately to see that no degradation exists. A following change should handle GPU/s.

1. https://bugzilla.redhat.com/1758330

File: tests/vmi_multus_test.go
Comment (alonSadan) [reply tests/vmi_multus_test.go:837]: Done.

> Maybe worth to mention the libvirt version where it was first not needed?

add to the commit message.

**Comment by alonSadan** [comment]
> Thank you! Great news.
> 
> See inline comments and below about the commit.
> 
> > sriov: remove host mounts for virt-launcher pods
> 
> tests, sriov: Do not mount `/sys/devices/` for SRIOV devices
> 
> > In the past, for sriov-vms, libvirt needed to open the `config` file under /sys/devices in
> > read-write mode. This was not possible since k8s mounted it as read only.
> > To overcome this sys/devices/ was mounted to the sriov virt-launcher pods in
> > read-write mode. Since then libvirt was patched to not attempt to open
> > the `config` file in read-write mode, and the whole /sys/* is mounted
> > as read-only to the virt-launcher pods. For that, there is no need for
> > the mount of /sys/devices anymore.
> 
> Libvirt required a Read/Write access to the hostdev devices representation on sysfs in older versions (< 5.6). Since libvirt 5.6.0-10 [1] this is no longer the case and Read(Only) access is sufficient.
> 
> Therefore, there is no longer a need for kubevirt to explicitly mount `/sys/devices` with RW access on the virt-launcher.
> 
> > This commit removes the mount of /sys/devices/ completely, modifies the unit test
> > and adds a functional test, to verify it.
> 
> This change removes the sysfs mount for VMI/s which contain SRIOV interfaces.
> 
> It does not change the logic for VMI/s which contain GPU devices as those need to be examined and tested separately to see that no degradation exists. A following change should handle GPU/s.
> 
> 1. https://bugzilla.redhat.com/1758330

Done.

File: go.mod
Comment (alonSadan) [reply go.mod]: removed.

**Comment by oshoval** [comment]
> Libvirt required a Read/Write access to the hostdev devices representation,
> on sysfs in older versions (< 5.6).
> Since libvirt 5.6.0-10 [1] this is no longer the case and Read(Only) access is sufficient.

Please either add [1] or remove the [1]
Please change to Read only

> Therefore, there is no longer a need for kubevirt to explicitly mount
> /sys/devices with RW access on the virt-launcher.
> 
> This change removes the sysfs mount for VMI/s which contain SRIOV interfaces.

Maybe worth to be explicit that it was /sys/devices mount (i see u mentioned it one line above, but sysfs is general)

> 
> It does not change the logic for VMI/s which contain GPU devices,
> as those need to be examined and tested separately to see that no degradation exists.
> A following change should handle GPU/s.
> 
> This commit removes the mount of /sys/devices/ completely, modifies the unit test
> and adds a functional test, to verify it.

Some of this info is outdated, some repeats, please refactor.

File: cmd/virt-handler/virt_launcher.cil
Comment (EdDev) [inline cmd/virt-handler/virt_launcher.cil:59]: I am not sure you can remove this as we still need to mount GPU devices at this stage. Maybe just change SR-IOV to GPU.

**Review by EdDev — COMMENTED** [review]
> Since libvirt 5.6.0-10 [1] this is no longer the case

You have not added the link to the BZ.
Either add it or remove the ref (`[1]`).

File: cmd/virt-handler/virt_launcher.cil
Comment (oshoval) [reply cmd/virt-handler/virt_launcher.cil:59]: maybe would worth to ask Jed or to try GPU out just in order to see if it affect it ?
git history can teach about it as well maybe, it was added after both 
SRIOV and GPU mounts were there and said SRIOV only.

Comment (EdDev) [reply cmd/virt-handler/virt_launcher.cil:59]: I think a simple common sense is enough here.
There is not difference really between the two devices in this context, both are passed into the VM in a similar manner (hostdev passthrough).

@alonSadan , you can send a separate PR to remove the GPU on top of this PR and remove this section as well.
It can then be reviewed by the relevant people.

Comment (alonSadan) [reply cmd/virt-handler/virt_launcher.cil]: OK. removed. I will follow up with a PR, add the link here.

**Comment by alonSadan** [comment]
> > Since libvirt 5.6.0-10 [1] this is no longer the case
> 
> You have not added the link to the BZ.
> Either add it or remove the ref (`[1]`).

added

**Comment by alonSadan** [comment]
> > Libvirt required a Read/Write access to the hostdev devices representation,
> > on sysfs in older versions (< 5.6).
> > Since libvirt 5.6.0-10 [1] this is no longer the case and Read(Only) access is sufficient.
> 
> Please either add [1] or remove the [1]
> Please change to Read only
> 
> > Therefore, there is no longer a need for kubevirt to explicitly mount
> > /sys/devices with RW access on the virt-launcher.
> > This change removes the sysfs mount for VMI/s which contain SRIOV interfaces.
> 
> Maybe worth to be explicit that it was /sys/devices mount (i see u mentioned it one line above, but sysfs is general)
> 
> > It does not change the logic for VMI/s which contain GPU devices,
> > as those need to be examined and tested separately to see that no degradation exists.
> > A following change should handle GPU/s.
> > This commit removes the mount of /sys/devices/ completely, modifies the unit test
> > and adds a functional test, to verify it.
> 
> Some of this info is outdated, some repeats, please refactor.

Done.

**Comment by alonSadan** [comment]
@oshoval  @EdDev  also note that I changed the unit test.

**Review by EdDev — APPROVED** [review]
Thank you!

**Review by oshoval — APPROVED** [review]
Thanks
nit for next time: imo its better to use [1] instead of 1. in the links list

**Comment by oshoval** [comment]
please add Release notes, even None, else it wont merge

**Review by ormergi — CHANGES_REQUESTED** [review]
HI Alon 
Could you please handle SELinux policy file this file as well?https://github.com/kubevirt/kubevirt/blob/master/cmd/virt-handler/virt_launcher.cil#L57

**Comment by oshoval** [comment]
> Hi Alon
> Could you please handle SELinux policy file this file as well?https://github.com/kubevirt/kubevirt/blob/master/cmd/virt-handler/virt_launcher.cil#L57

See please
https://github.com/kubevirt/kubevirt/pull/4475#discussion_r515755848

**Comment by ormergi** [comment]
> Hi Alon
> Could you please handle SELinux policy file this file as well?https://github.com/kubevirt/kubevirt/blob/master/cmd/virt-handler/virt_launcher.cil#L57

I just saw the discussion about it sorry
/lgtm

File: cmd/virt-handler/virt_launcher.cil
Comment (ormergi) [reply cmd/virt-handler/virt_launcher.cil]: Since there is no realy difference for those two devices in this context I think this should be addressed in this PR.
Why do we need to worry about it so much there is a comment the explicitly says its for SRIOV only?
If it was used for GPU I would expect for another comment.
Do we have a lane for GPU?

Comment (oshoval) [reply cmd/virt-handler/virt_launcher.cil]: code still mounts sysfs dir for GPU, code cant be removed without tested, either you test it or you dont touch it and doing it on a follow PR.

Comment (ormergi) [reply cmd/virt-handler/virt_launcher.cil]: You dont know GPU used it, you assume it is.
What will be the difference to remove it in another PR?

Comment (oshoval) [reply cmd/virt-handler/virt_launcher.cil]: and you dont know GPU doesnt need it, you assume it is.
see please all the discussion above, we dont remove code just because we assume something without testing it.

Comment (alonSadan) [reply cmd/virt-handler/virt_launcher.cil]: draft PR for now: 
https://github.com/kubevirt/kubevirt/pull/4476

Comment (ormergi) [reply cmd/virt-handler/virt_launcher.cil]: I dont think we have a lane to test it

Comment (alonSadan) [reply cmd/virt-handler/virt_launcher.cil]: > I dont think we have a lane to test it

I'll keep it draft then until it is verified

**Comment by alonSadan** [comment]
/hold 
let's verify that sriov lane passes on prow
/test pull-kubevirt-e2e-kind-1.17-sriov

**Comment by alonSadan** [comment]
/unhold
CI passed

**Pull request merged**
